library/std/src/sys/unix/futex.rs

```diff
@@ -0,0 +1,37 @@
+#![cfg(any(target_os = "linux", target_os = "android"))]
+
+use crate::convert::TryInto;
+use crate::ptr::null;
+use crate::sync::atomic::AtomicI32;
+use crate::time::Duration;
+
+pub fn futex_wait(futex: &AtomicI32, expected: i32, timeout: Option<Duration>) {
+    let timespec = timeout.and_then(|d| {
+        Some(libc::timespec {
+            // Sleep forever if the timeout is longer than fits in a timespec.
```

Comment:
You could also saturate to the maximum timeout. `thread::park` is allowed to spuriously wake up.

Reply:
i32::MAX seconds is more than 68 years, and most systems use a i64 (2e+11 years) here, so I don't think it really matters. (Note this is a duration, not an absolute timestamp.) The implementations I've seen round *up* (in case the underlying call doesn't do nanosecond precision, e.g. on Windows), and overflow to infinity. So I just followed that.